cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -414,9 +415,15 @@ public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
                                          "The following programs are still running: " + appAllRunningPrograms);
     }
     //All Apps are STOPPED, delete them
+    Set<String> appNames = new HashSet<>();
     for (ApplicationSpecification appSpec : allSpecs) {
-      ApplicationId id = new ApplicationId(namespaceId.getId(), appSpec.getName());
-      deleteApp(id, appSpec);
+      appNames.add(appSpec.getName());
+    }
+    for (String appName : appNames) {
+      Collection<ApplicationId> appIds = store.getAllAppVersionsAppIds(namespace, appName);
```

Comment:
`store.getAllApplications(namespaceId.toEntityId())` should return all versions of all applications, right? Do we still need to do this extra step?